jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/Utils.scala

```diff
@@ -14,19 +14,57 @@
  limitations under the License.
  */
 
-package ml.dmlc.xgboost4j.scala.spark.util
+package ml.dmlc.xgboost4j.scala.spark
 
+import org.apache.spark.ml.feature.{LabeledPoint => MLLabeledPoint}
+import org.apache.spark.ml.linalg.{DenseVector, SparseVector, Vector, Vectors}
 import org.json4s.{DefaultFormats, FullTypeHints, JField, JValue, NoTypeHints, TypeHints}
 
-// based on org.apache.spark.util copy /paste
-object Utils {
+import ml.dmlc.xgboost4j.{LabeledPoint => XGBLabeledPoint}
+
+private[scala] object Utils {
+
+  private[spark] implicit class XGBLabeledPointFeatures(
+      val labeledPoint: XGBLabeledPoint
+  ) extends AnyVal {
+    /** Converts the point to [[MLLabeledPoint]]. */
+    private[spark] def asML: MLLabeledPoint = {
+      MLLabeledPoint(labeledPoint.label, labeledPoint.features)
+    }
+
+    /**
+     * Returns feature of the point as [[org.apache.spark.ml.linalg.Vector]].
+     */
+    def features: Vector = if (labeledPoint.indices == null) {
+      Vectors.dense(labeledPoint.values.map(_.toDouble))
+    } else {
+      Vectors.sparse(labeledPoint.size, labeledPoint.indices, labeledPoint.values.map(_.toDouble))
+    }
+  }
+
+  private[spark] implicit class MLVectorToXGBLabeledPoint(val v: Vector) extends AnyVal {
+    /**
+     * Converts a [[Vector]] to a data point with a dummy label.
+     *
+     * This is needed for constructing a [[ml.dmlc.xgboost4j.scala.DMatrix]]
+     * for prediction.
+     */
+    // TODO support sparsevector
```

Comment:
I see `SparseVector` being matched in the next statement, is it not supported?

Reply:
The SparseVector is still supported, and it will be converted to dense to build DMatrix

Reply:
Is the missing value being filled with 0s? If so, what happens if the 0s indeed represent missing values? Also, this is a breaking change I believe, which should be part of the migration guide.

Reply:
No exactly, missing value can be any value including 0. The sparsevector will be converted to densevector, and the missing value will be passed to the xgboost when constructing the DMatrix.

Reply:
Yes, during the conversion from sparse to dense vector, what value is being filled into the missing entries of the input sparse vector?

Reply:
Done.